### ISSUE TITLE
Security validation incorrect

### ISSUE THREAD

**Issue by sgovaerts** [readysitego/spgo#64]
This one is driving me crazy, I was able to update a file once but not after that. While using NTML authentication on mixed authentication (NTML + FBA) it gets access denied for my own user which I can use to upload & overwrite the files manually but not using SPGO. If I use a site administrator user, I keep getting the "The security validation for this page is invalid.", I was able to update a file once and then after restarting visual studio code it fails.

````
    10/26/2018 08:40:18.89	w3wp.exe (0xAA94)	0x25EC	SharePoint Foundation	Monitoring	nasq	Medium	Entering monitored scope (Request (GET:https://site.domain.com:443/sites/demo/_api/web/GetFileByServerRelativeUrl('/sites/demo/SiteAssets/printOnClick_mr.js')/CheckOut())). Parent No	35129c9e-acfd-3093-e956-61d513561ab1    
    10/26/2018 08:40:18.89	w3wp.exe (0xAA94)	0x25EC	SharePoint Foundation	Logging Correlation Data	xmnv	Medium	Name=Request (GET:https://site.domain.com:443/sites/demo/_api/web/GetFileByServerRelativeUrl('/sites/demo/SiteAssets/printOnClick_mr.js')/CheckOut())	6f149c9e-bcfc-3093-e956-69fe7a3a4844    
    10/26/2018 08:40:18.92	w3wp.exe (0xAA94)	0x5ED0	SharePoint Foundation	CSOM	agw10	Medium	Begin CSOM Request ManagedThreadId=189, NativeThreadId=24272	6f149c9e-bcfc-3093-e956-6f34a799ceba    
    10/26/2018 08:40:18.92	w3wp.exe (0xAA94)	0x5ED0	SharePoint Foundation	Logging Correlation Data	xmnv	Medium	Site=/sites/demo	6f149c9e-bcfc-3093-e956-6f34a799ceba    
    10/26/2018 08:40:18.92	w3wp.exe (0xAA94)	0x5ED0	SharePoint Foundation	General	8kh7	High	The security validation for this page is invalid. Click Back in your Web browser, refresh the page, and try your operation again.	6f149c9e-bcfc-3093-e956-6f34a799ceba    
    10/26/2018 08:40:18.92	w3wp.exe (0xAA94)	0x5ED0	SharePoint Foundation	General	aix9j	High	SPRequest.ValidateFormDigest: UserPrincipalName=i:0).w|, AppPrincipalName= ,bstrUrl=https://site.domain.com/sites/demo ,bstrListName=<null>	6f149c9e-bcfc-3093-e956-6f34a799ceba    
    10/26/2018 08:40:18.92	w3wp.exe (0xAA94)	0x5ED0	SharePoint Foundation	General	ai1wu	Medium	System.Runtime.InteropServices.COMException: The security validation for this page is invalid. Click Back in your Web browser, refresh the page, and try your operation again., StackTrace:    at Microsoft.SharePoint.SPWeb.ValidateFormDigest()
    10/26/2018 08:40:18.92	w3wp.exe (0xAA94)	0x5ED0	SharePoint Foundation	CSOM	afxwb	High	System.Runtime.InteropServices.COMException (0x8102006D): The security validation for this page is invalid. Click Back in your Web browser, refresh the page, and try your operation again.     at Microsoft.SharePoint.Library.SPRequestInternalClass.ValidateFormDigest(String bstrUrl, String bstrListName)
    10/26/2018 08:40:18.92	w3wp.exe (0xAA94)	0x5ED0	SharePoint Foundation	CSOM	agmjp	High	Original error: System.Runtime.InteropServices.COMException (0x8102006D): The security validation for this page is invalid. Click Back in your Web browser, refresh the page, and try your operation again.     at Microsoft.SharePoint.Library.SPRequestInternalClass.ValidateFormDigest(String bstrUrl, String bstrListName)
    10/26/2018 08:40:18.92	w3wp.exe (0xAA94)	0x5ED0	SharePoint Portal Server	Microfeeds	aizmo	Medium	SocialRESTExceptionProcessingHandler.DoServerExceptionProcessing - SharePoint Server Exception [System.Runtime.InteropServices.COMException (0x8102006D): The security validation for this page is invalid. Click Back in your Web browser, refresh the page, and try your operation again.     at Microsoft.SharePoint.Library.SPRequestInternalClass.ValidateFormDigest(String bstrUrl, String bstrListName)
    10/26/2018 08:40:18.92	w3wp.exe (0xAA94)	0x1370	SharePoint Foundation	Monitoring	b4ly	Medium	Leaving Monitored Scope (Request (POST:https://site.domain.com:443/sites/demo/_api/web/GetFileByServerRelativeUrl('/sites/demo/SiteAssets/printOnClick_mr.js')/CheckOut())). Execution Time=38,1100810298516	6f149c9e-bcfc-3093-e956-6f34a799ceba    
````

Any idea? Weirdest part is that I could update the file once before but can't replicate it anymore.
Many thanks

**Comment by koltyakov** [comment]
@sg-chrishasz, we went through this before, for mixed auth cases the `'x-forms_based_auth_accepted': 'f'` header has to be in the requests. [Details](https://github.com/s-KaiNet/node-sp-auth/issues/54).

**Comment by sgovaerts** [comment]
@koltyakov but shouldn't line 74 etc take care of that? https://github.com/readysitego/spgo/blob/master/src/util/requestHelper.ts

**Comment by koltyakov** [comment]
Looks like it is. Didn't know that it was already there. Well, so we wait for Chris's suggestion.

**Comment by sgovaerts** [comment]
@koltyakov Just checked and you were right, it's missing for the createRequest.

    static createRequest(appManager) {
        process.env['_sp_request_headers'] = JSON.stringify({
            'X-FORMS_BASED_AUTH_ACCEPTED': 'f'
        });
        return spRequest.create(this.createCredentials(appManager));
    }

After adding that code it works! Thank you for the help.
@sg-chrishasz can you add the header also to the createRequest, otherwise it doesn't work in mixed authenticatione environments.

**Comment by s-KaiNet** [comment]
It seems that for some reason [this line](https://github.com/readysitego/spgo/blob/master/src/service/spFileService.ts#L140) isn't get executed. This line sets `X-FORMS_BASED_AUTH_ACCEPTED` header.

**Comment by sg-chrishasz** [comment]
@s-KaiNet - As always I appreciate your assistance - you are spot-on with the root cause!

@sgovaerts - great catch! You are seeing this issue because I am not consistently setting the `'X-FORMS_BASED_AUTH_ACCEPTED': 'f'` header every time I create an HTTP request to SharePoint. 

**Moment of humility:** I even have a `//TODO` in the code to refactor how I manage HTTP headers for NTLM auth. 

I'm going to refactor this in my next SPGo release, which I promise to push this weekend.

In the short-term, if you want to patch this issue in your local copy I would recommend you replace the `verify` function - lines 80 to 100 in `authenticationService.ts` with the version below (as you suggested above):

``` typescript
		// Test the new credential set to make sure it is valid
		function verify(appManager){
			return new Promise(function(resolve, reject) {
				
				let spr = RequestHelper.createRequest(appManager);

				//TODO: come up with a more consistent solution for setting headers
				if( Constants.SECURITY_NTLM == vscode.window.spgo.config.authenticationType){
					RequestHelper.setNtlmHeader();
				}

				spr.requestDigest(vscode.window.spgo.config.sharePointSiteUrl)
					.then(function(){ //response => {
						//store credentials?
						if(appManager.config.storeCredentials){
							CredentialDao.setCredentials(vscode.window.spgo.config.sharePointSiteUrl, appManager.credentials);
							return processNextCommand(appManager);
						}
						resolve(appManager);
					}, err => {
						appManager.credentials = null;
						reject(err);
					});
			});
		}
```

Glad you enjoy SPGo, thanks for your use and support!

**Comment by sg-chrishasz** [comment]
Resolved in release 1.3.3
